packages/server/graphql/traceGraphQL.ts

```diff
@@ -0,0 +1,357 @@
+/**
+ * @file Wrapping functions to instrument compiled queries.
+ * These conflict with the dd-trace graphql plugin since the use the same fields to store intermediate data
+ */
+import {Span, SpanOptions, TraceOptions, Tracer} from 'dd-trace'
+import {
+  DefinitionNode,
+  DocumentNode,
+  ExecutionResult,
+  GraphQLField,
+  GraphQLFieldResolver,
+  GraphQLObjectType,
+  GraphQLOutputType,
+  GraphQLResolveInfo,
+  GraphQLSchema,
+  OperationDefinitionNode
+} from 'graphql'
+import {Path} from 'graphql/jsutils/Path'
+import {CompiledQuery, compileQuery, CompilerOptions, isCompiledQuery} from 'graphql-jit'
+
+interface ExecutionArgs {
+  rootValue?: any
+  contextValue?: any
+  variableValues?: any
+}
+
+interface Config {
+  /**
+   * Whether to collapse list items into a single element. (i.e. single
+   * `users.*.name` span instead of `users.0.name`, `users.1.name`, etc)
+   *
+   * @default true
+   */
+  collapse?: boolean
+
+  /**
+   * The maximum depth of fields/resolvers to instrument. Set to `0` to only
+   * instrument the operation or to `-1` to instrument all fields/resolvers.
+   *
+   * @default -1
+   */
+  depth?: number
+
+  /**
+   * An object of optional callbacks to be executed during the respective
+   * phase of a GraphQL operation. Undefined callbacks default to a noop
+   * function.
+   *
+   * @default {}
+   */
+  hooks?: {
+    execute?: (span: Span, args: ExecutionArgs, res?: any) => void
+  }
+}
+
+/**
+ * Return {@link compileQuery} function which will add tracing to the query.
+ */
+export function tracedCompileQuery(tracer: Tracer, config: Config) {
+  if (process.env.DD_TRACE_ENABLED !== 'true') {
+    return compileQuery
+  }
+  return function compileTracedQuery(
+    schema: GraphQLSchema,
+    document: DocumentNode,
+    operationName?: string,
+    partialOptions?: Partial<CompilerOptions>
+  ) {
+    wrapSchema(tracer, config, schema)
+    const query = compileQuery(schema, document, operationName, partialOptions)
+    if (isCompiledQuery(query)) {
+      const operation = getOperation(document, operationName)
```

Comment:
having getOperation return `OperationDefinitionNode | undefined` is correct. 
however, in this specific case, i'd do a nonnull coercion inside this conditional because if the query compiled, then we are guaranteed that there is one matching definition & then we can get rid of 1 of the optional chains


```ts
      const operation = getOperation(document, operationName)!
```